common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java

```diff
@@ -129,9 +133,31 @@ public static void setPerfLogger(PerfLogger resetPerfLogger) {
    * @param method method or ID that identifies this perf log element.
    */
   public void perfLogBegin(String callerName, String method) {
+    perfLogBegin(callerName, method, null);
+  }
+
+  /**
+   * Call this function when you start to measure time spent by a piece of code.
+   * @param callerName the logging object to be used.
+   * @param method method or ID that identifies this perf log element.
+   * @param additionalInfo particular step within the method that is being measured
+   */
+  public void perfLogBegin(String callerName, String method, String additionalInfo) {
     long startTime = System.currentTimeMillis();
-    startTimes.put(method, Long.valueOf(startTime));
-    LOG.debug("<PERFLOG method={} from={}>", method, callerName);
+    String key = method;
+    if (additionalInfo != null) {
+      key = key + " - " + additionalInfo;
```

Comment:
It doesn't seem that `additionalInfo` was ever meant to be used as a key. Instead of concatenating the key with the info in order to obtain a unique id maybe it would be better that the caller passes the appropriate measurement event in the first place. This class has many static constants for the various events that we want to track but as far as I understand we want to have more fine grained measurements. 

For instance, it seems that we want to track `PerfLogger.OPTIMIZER` in more detail. If that's the case then possibly it makes sense to introduce more entries for the various sub-steps.

Reply:
Makes sense, I will do this change.